File: 0147-insertion-sort-list/0147-insertion-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        
        if(head == NULL || head->next == NULL) return head;
        
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        
        ListNode* prev = head, *cur = head->next;
        while(cur != NULL){
            
            if(cur->val >= prev->val){
                prev = cur;
                cur = cur->next;
                continue;
            }
            
            ListNode* tmp = dummy;
            while(cur->val > tmp->next->val) tmp = tmp->next;
            
            prev->next = cur->next;
            cur->next = tmp->next;
            tmp->next = cur;
            cur = prev->next;
        }
        
        return dummy->next;
    }
};

File: 0147-insertion-sort-list/0147-insertion-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        
        if(head == null || head.next == null) return head;
        
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        
        ListNode prev = head, cur = head.next;
        while(cur != null){
            if(cur.val >= prev.val){
                prev = cur;
                cur = cur.next;
                continue;
            }
            
            ListNode temp = dummy;
            while(cur.val > temp.next.val) temp = temp.next;
            
            prev.next = cur.next;
            cur.next = temp.next;
            temp.next = cur;
            cur = prev.next;
        }
        
        return dummy.next;
    }
}